### DISCUSSION TITLE
Modified Phalcon\Di\Injectable - Remove unused Varies :))

### DISCUSSION THREAD

**Discussion by Sosiashvili** [phalcon/phalcon#205] [Ideas]
I Made BaseModel.php 

Class BaseModel extends \Phalcon\Di\Injectable {
}

I just modified this function: :::::  public function __get($propertyName)

//Fallback to the PHP userland if the cache is not available
if ($dependencyInjector->has($propertyName) === true) {
     $service = $dependencyInjector->getShared($propertyName);
     **$this->$propertyName = $service;** << I removed this string
      return $service;
}

what happens if remove this???

child class works as to before.

**Comment by borisdelev** [comment]
... i think u didnt need that fallback at all. If that BaseModel is a Model wrapper or something, need to extends \Phalcon\Mvc\Model - in that case u didnt have $this->di (must get static: ```$di = \Phalcon\Di::getDefault()```).... and its rare to use shared service in Models.

So depends what is that BaseModel and what function do and for what reason u need.

Anyway- in every class who extends Di... you have $this->has('...') and every added shared service.

**Reply by Sosiashvili** [reply]
Thanks **Boris**

I want to use vars in the module...

But when I use: $this->db or $this->filter and when this class is in var_dump take a lot vars in it...

why is this so necessary that these services vars show in models??

**Reply by Sosiashvili** [reply]
When I made this trick my model looks like I wanted it...

but __get function to have bad habit by this situation...

when undefined var is called:  
$this->update->errerewrwrewr in Controller
Notice: Undefined property: Backend\Models\product\update::$errerewrwrewr in /home/core/Phalcon/Injectable.php on line 39

this var is used in the controller. by this moment I cant find where is this var used... because this text me that this is in Injectable.php :((

**Reply by Sosiashvili** [reply]
when I am using the original abstract class Phalcon\Di\Injectable
this error shows that is in the controller.

**Reply by Sosiashvili** [reply]
Notice: Access to undefined property errerewrwrewr in /home/apps/backend/controllers/ProductController.php on line 24

**Comment by Sosiashvili** [comment]
I want use services in model or any class but not show in it...

**Comment by niden** [comment]
> I Made BaseModel.php
> 
> Class BaseModel extends \Phalcon\Di\Injectable { }
> 
> I just modified this function: ::::: public function __get($propertyName)
> 
> //Fallback to the PHP userland if the cache is not available if ($dependencyInjector->has($propertyName) === true) { $service = $dependencyInjector->getShared($propertyName); **$this->$propertyName = $service;** << I removed this string return $service; }
> 
> what happens if remove this???
> 
> child class works as to before.

You can certainly remove it if you do not want that. However some functionality will change on the DI. For instance you will not be able to access services with their names as properties. Example:

```php
$value = $this->response->getPost('email');
```

you will have to change that code to something like this:
```php
$value = $this->getDI()->get('response')->getPost('email');
```
or
```php
$value = $this->getResponse()->getPost('email');
```


If you are fine with that then go for it. 

`__get` was implemented years ago in the DI to help developers access services quickly as properties in injectable components such as controllers. Although I do not agree with its use or the other deviations that it introduced, it is there and if it is removed it will cause a lot of issues with existing applications - pretty much rendering them useless because all the related code will have to be refactored.

In v6 we will introduce a much more flexible DI container which will incorporate lazy loading, PSR-11 as well as automatic injection of services.
